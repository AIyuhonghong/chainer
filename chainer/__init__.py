import pkg_resources

from chainer import function
from chainer import function_set
from chainer.functions import basic_math
from chainer import optimizer
from chainer import serializer
from chainer import variable


__version__ = pkg_resources.get_distribution('chainer').version

AbstractSerializer = chainer.AbstractSerializer
Deserializer = chainer.Deserializer
Function = function.Function
FunctionSet = function_set.FunctionSet
Optimizer = optimizer.Optimizer
Serializer = chainer.Serializer
Variable = variable.Variable

basic_math.install_variable_arithmetics()
