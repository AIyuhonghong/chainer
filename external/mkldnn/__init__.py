from external.mkldnn import prepare_mkldnn  # NOQA
from external.mkldnn.prepare_mkldnn import prepare  # NOQA
from external.mkldnn.prepare_mkldnn import root  # NOQA
from external.mkldnn.prepare_mkldnn import lib_path  # NOQA
from external.mkldnn.prepare_mkldnn import include_path  # NOQA
from external.mkldnn.prepare_mkldnn import source_path  # NOQA
