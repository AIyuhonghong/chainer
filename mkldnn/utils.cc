/*
 *COPYRIGHT
 *All modification made by Intel Corporation: Â© 2017 Intel Corporation.
 *Copyright (c) 2015 Preferred Infrastructure, Inc.
 *Copyright (c) 2015 Preferred Networks, Inc.
 *
 *Permission is hereby granted, free of charge, to any person obtaining a copy
 *of this software and associated documentation files (the "Software"), to deal
 *in the Software without restriction, including without limitation the rights
 *to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 *copies of the Software, and to permit persons to whom the Software is
 *furnished to do so, subject to the following conditions:
 *
 *The above copyright notice and this permission notice shall be included in
 *all copies or substantial portions of the Software.
 *
 *THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 *IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 *FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 *AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 *LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 *OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 *THE SOFTWARE.
 *
 *
 *######################################################################
 *# The CuPy is designed based on NumPy's API.
 *# CuPy's source code and documents contain the original NumPy ones.
 *######################################################################
 *Copyright (c) 2005-2016, NumPy Developers.
 *All rights reserved.
 *
 *Redistribution and use in source and binary forms, with or without
 *modification, are permitted provided that the following conditions are
 *met:
 *
 *    * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *    * Redistributions in binary form must reproduce the above
 *       copyright notice, this list of conditions and the following
 *       disclaimer in the documentation and/or other materials provided
 *       with the distribution.
 *
 *    * Neither the name of the NumPy Developers nor the names of any
 *       contributors may be used to endorse or promote products derived
 *       from this software without specific prior written permission.
 *
 *THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 *A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 *OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 *SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 *LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 *DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 *THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 *(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 *OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *######################################################################
 */


#if defined(__INTEL_COMPILER) && (__INTEL_COMPILER >= 1300)
#include <immintrin.h>
int has_intel_knl_features()
{
  const unsigned long knl_features =
      (_FEATURE_AVX512F | _FEATURE_AVX512ER |
       _FEATURE_AVX512PF | _FEATURE_AVX512CD );
  return _may_i_use_cpu_feature( knl_features );
}

int has_intel_avx2_features()
{
  const unsigned long avx2_features = _FEATURE_AVX2;
  return _may_i_use_cpu_feature( avx2_features );
}
#else /* non-Intel compiler */
#include <stdint.h>
#if defined(_MSC_VER)
#include <intrin.h>
#endif
void run_cpuid(uint32_t eax, uint32_t ecx, uint32_t* abcd)
{
#if defined(_MSC_VER)
  __cpuidex(abcd, eax, ecx);
#else
  uint32_t ebx = 0, edx = 0;
 #if defined( __i386__ ) && defined ( __PIC__ )
  /* in case of PIC under 32-bit EBX cannot be clobbered */
  __asm__ ( "movl %%ebx, %%edi \n\t cpuid \n\t xchgl %%ebx, %%edi" : "=D" (ebx),
 # else
  __asm__ ( "cpuid" : "+b" (ebx),
 # endif
              "+a" (eax), "+c" (ecx), "=d" (edx) );
        abcd[0] = eax; abcd[1] = ebx; abcd[2] = ecx; abcd[3] = edx;
#endif
}

int check_xcr0_zmm() {
  uint32_t xcr0;
  uint32_t zmm_ymm_xmm = (7 << 5) | (1 << 2) | (1 << 1);
#if defined(_MSC_VER)
  xcr0 = (uint32_t)_xgetbv(0);  /* min VS2010 SP1 compiler is required */
#else
  __asm__ ("xgetbv" : "=a" (xcr0) : "c" (0) : "%edx" );
#endif
  return ((xcr0 & zmm_ymm_xmm) == zmm_ymm_xmm); /* check if xmm, zmm and zmm state are enabled in XCR0 */
}

int check_xcr0_ymm()
{
 uint32_t xcr0;
  uint32_t ymm_xmm = (1 << 2) | (1 << 1);
#if defined(_MSC_VER)
 xcr0 = (uint32_t)_xgetbv(0); /* min VS2010 SP1 compiler is required */
#else
 __asm__ ("xgetbv" : "=a" (xcr0) : "c" (0) : "%edx" );
#endif
 return ((xcr0 & ymm_xmm) == ymm_xmm); /* checking if xmm and ymm state are enabled in XCR0 */
}

int has_intel_knl_features() {
    uint32_t abcd[4];
    uint32_t osxsave_mask   =   (1 << 27); // OSX.
    uint32_t knl_bmi12_mask =   (1 << 16) | // AVX-512F
                                (1 << 26) | // AVX-512PF
                                (1 << 27) | // AVX-512ER
                                (1 << 28);  // AVX-512CD
    run_cpuid( 1, 0, abcd );
    // step 1 - must ensure OS supports extended processor state management
    if ( (abcd[2] & osxsave_mask) != osxsave_mask )
        return 0;
    run_cpuid( 7, 0, abcd );
    if ( (abcd[1] & knl_bmi12_mask) != knl_bmi12_mask )
        return 0;
    // step 2 - must ensure OS supports ZMM registers (and YMM, and XMM)
    if ( ! check_xcr0_zmm() )
        return 0;

    return 1;
}

int has_intel_avx2_features() {
    uint32_t abcd[4];
    uint32_t fma_movbe_osxsave_mask = ((1 << 12) | (1 << 22) | (1 << 27));
    uint32_t avx2_bmi12_mask = (1 << 5) | (1 << 3) | (1 << 8);
    run_cpuid( 1, 0, abcd );
    if ( (abcd[2] & fma_movbe_osxsave_mask) != fma_movbe_osxsave_mask )
        return 0;
    if ( ! check_xcr0_ymm() )
        return 0;
    run_cpuid( 7, 0, abcd );
    if ( (abcd[1] & avx2_bmi12_mask) != avx2_bmi12_mask )
        return 0;
    return 1;
}
#endif /* non-Intel compiler */

static int can_use_intel_knl_features() {
  static int knl_features_available = -1;
  /* test is performed once */
  if (knl_features_available < 0 )
    knl_features_available = has_intel_knl_features();
  return knl_features_available;
}

static int can_use_intel_avx2_features() {
  static int avx2_features_available = -1;
  /* test is performed once */
  if (avx2_features_available < 0 )
    avx2_features_available = has_intel_avx2_features();
  return avx2_features_available;
}

int cpu_support_avx512_p()
{
    return can_use_intel_knl_features();
    // TODO the following code requires gcc 6 or above
    //__builtin_cpu_init();
    //return  true && __builtin_cpu_supports("avx512f")
    //             && __builtin_cpu_supports("avx512cd")
    //             && __builtin_cpu_supports("avx512er")
    //             && __builtin_cpu_supports("avx512pf");
    // TODO icc path not enabled
    //const unsigned long knl_features =
    //    (_FEATURE_AVX512F | _FEATURE_AVX512ER |
    //    _FEATURE_AVX512PF | _FEATURE_AVX512CD );
    //    return _may_i_use_cpu_feature( knl_features );
}

int cpu_support_avx2_p()
{
    return can_use_intel_avx2_features();
    //__builtin_cpu_init();
    //return 1 && __builtin_cpu_supports("avx2");
    // TODO icc path not enabled
    //const unsigned long avx2_features = _FEATURE_AVX2
    //    return _may_i_use_cpu_feature( avx2_features );
}


