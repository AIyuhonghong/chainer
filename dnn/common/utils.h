#ifndef _UTILS_H_
#define _UTILS_H_

#include <glog/logging.h>
#include <mkldnn.hpp>
#include <iostream>
#include "op_param.h"
#include "omp.h"
using namespace mkldnn;

memory::format get_desired_format(int channel);

template<typename T>
void eltwise_multiply(T* x1, T* x2, T* y, size_t n) {
#pragma omp parallel for schedule(static)
    for (size_t i = 0; i < n; ++i) {
        y[i] = x1[i] * x2[i];
    }
}

//
//// map C type with mkldnn's
//// float -> memory::data_type::f32
//// int -> memory::data_type::s32
//// int16_t -> memory::data_type::s16
//// int8_t -> memory::data_type::s8
//// uint8_t -> memory::data_type::u8
//
template<typename T>
static inline mkldnn::memory::data_type memory_data_type() {
    if (typeid(T) == typeid(float))
        return mkldnn::memory::data_type::f32;
    else if (typeid(T) == typeid(int))
        return mkldnn::memory::data_type::s32;
    else if (typeid(T) == typeid(int16_t))
        return mkldnn::memory::data_type::s16;
    else if (typeid(T) == typeid(int8_t))
        return mkldnn::memory::data_type::s8;
    else if (typeid(T) == typeid(uint8_t))
        return mkldnn::memory::data_type::u8;

    LOG(ERROR) << "Not support type";
    return mkldnn::memory::data_type::data_undef;
}

// utils function conver int/double/bool/dims/ to string
static inline std::string int_to_string(int value) {
    std::ostringstream os;
    os << std::hex << "I" << value << "_";
    return os.str();
}

static inline std::string double_to_string(double value) {
    std::ostringstream os;
    os << "D" << value << "_";
    return os.str();
}

static inline std::string float_to_string(float value) {
    std::ostringstream os;
    os << "F" << value << "_";
    return os.str();
}

static inline std::string bool_to_string(bool value) {
    std::ostringstream os;
    os << "B" << value << "_";
    return os.str();
}

static inline std::string dims_to_string(mkldnn::memory::dims dims) {
   std::ostringstream os;
   os << "DIMS:";
   for (int i = 0; i < dims.size(); i++)
       os << dims[i] << ",";
   os << ";";
   return os.str();
}

static inline std::string long_to_string(size_t value) {
    std::ostringstream os;
    os << std::hex << "L" << value << "_";
    return os.str();
}

static inline mkldnn::algorithm pooling_algo_convert(pooling_param_t::algorithm input) {
    switch(input) {
        case pooling_param_t::algorithm::pooling_max:
            return mkldnn::pooling_max;
        case pooling_param_t::algorithm::pooling_avg:
            return mkldnn::pooling_avg;
        case pooling_param_t::algorithm::pooling_avg_include_padding:
            return mkldnn::pooling_avg_include_padding;
        case pooling_param_t::algorithm::pooling_avg_exclude_padding:
            return mkldnn::pooling_avg_exclude_padding;
        default:
            LOG(ERROR) << "Not a valid pooling algo";
            return mkldnn::pooling_max;
    }
}

static inline mkldnn::algorithm lrn_algo_convert(lrn_param_t::algorithm input) {
    switch(input) {
        case lrn_param_t::algorithm::lrn_across_channels:
            return mkldnn::lrn_across_channels;
        case lrn_param_t::algorithm::lrn_within_channel:
            return mkldnn::lrn_within_channel;
        default:
            LOG(ERROR) << "Not a valid lrn algo";
            return mkldnn::lrn_across_channels;
    }
}

template<typename T, typename U>
inline T div_up(const T a, const U b) {
    assert(b);
    return(a + b - 1) / b;
}
template <typename T, typename U>
inline void balance211(T n, U team, U tid, T &n_start, T &n_end) {
    T n_min = 1;
    T &n_my = n_end;
    if (team <= 1 || n == 0) {
        n_start = 0;
        n_my = n;
    } else if (n_min == 1) {
        // team = T1 + T2
        // n = T1*n1 + T2*n2  (n1 - n2 = 1)
        T n1 = div_up(n, (T)team);
        T n2 = n1 - 1;
        T T1 = n - n2 * (T)team;
        n_my = (T)tid < T1 ? n1 : n2; 
        n_start = (T)tid <= T1 ? tid * n1 : T1 * n1 + ((T)tid - T1) * n2; 
    }   

    n_end += n_start;
}

inline void fast_memcpy(char* data_o, char *data_i, int len)
{
    size_t nelems_float = len / 4;
    size_t nelems_char = len % 4;
    const int block_size = 16; 
    const auto num_blocks_float = nelems_float / block_size;
    const auto rem_elems_float =  nelems_float % block_size;
    float* output_f = (float*)data_o;
    float* input_f = (float*) data_i;
    char* output_c = (char*) data_o;
    char* input_c = (char*) data_i;
#   pragma omp parallel
    {   
        const int ithr = omp_get_thread_num();
        const int nthr = omp_get_num_threads();
        size_t start{0}, end{0};
        balance211(num_blocks_float, nthr, ithr, start, end);  
        start = start * block_size;
        end = end * block_size;
#       pragma omp simd
        for (size_t e = start; e < end; ++e) {
            output_f[e] = input_f[e];
        }   
        if (rem_elems_float != 0 && ithr ==  nthr -1 )  {
            for (int e = nelems_float - rem_elems_float; e < nelems_float; ++e) {
                output_f[e] = input_f[e];
            }   
        }   
        if (nelems_char != 0 && ithr ==  nthr -1){
            for (int e = nelems_float*4; e < len; ++e) {
                output_c[e] = input_c[e];
            }   
        }   
    }   
    return;
}

#endif // _UTILS_H_
