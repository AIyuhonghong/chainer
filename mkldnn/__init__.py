import sys
import ctypes

# For C++ extension to work
sys.setdlopenflags(sys.getdlopenflags() | ctypes.RTLD_GLOBAL)

# API lift
from mkldnn import api
from mkldnn import chainer
from mkldnn import compute_complex

from mkldnn.mdarray import mdarray

from mkldnn.chainer.fanout import *

from mkldnn.config import global_config  # NOQA
global_config.gx_opt = True
