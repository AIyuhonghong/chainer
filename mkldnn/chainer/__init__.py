import os
import numpy

from mkldnn.mdarray import mdarray

from chainer import variable
from chainer import configuration

from chainer.configuration import config
from chainer.configuration import global_config

# export CHAINER_ENABLE_COSIM=0
global_config.cosim = bool(int(os.environ.get('CHAINER_ENABLE_COSIM', '0')))
# export CHAINER_ENABLE_COSIM_CONTINUE=0
global_config.cosim_continue = bool(int(os.environ.get('CHAINER_ENABLE_COSIM_CONTINUE', '0')))

if global_config.cosim is True:
    variable.RANK_START = 1


def is_cosim_continue():
    """Get the cosim continue mode

    Returns:
        bool: Return ``True`` if continue after a failure.
    """
    return config.cosim_continue


def is_cosim():
    """Get the cosim mode.

    Returns:
        bool: Return ``True`` if Chainer is in cosim mode.
    """
    return config.cosim


def plain_array(params):
    assert isinstance(params, tuple) \
            or isinstance(params, list) \
            or isinstance(params, mdarray) \
            or isinstance(params, numpy.ndarray) \
            or isinstance(params, variable.Variable)

    _params = ()

    if isinstance(params, variable.Variable):
        return (numpy.array(params.data), )
    elif isinstance(params, numpy.ndarray):
        return (params, )
    elif isinstance(params, mdarray):
        return (numpy.array(params), )

    for p in params:
        if isinstance(p, variable.Variable):
            p = numpy.array(p.data)
        if isinstance(p, mdarray):
            _params += (numpy.array(p), )
        else:
            _params += (p, )

    return _params

