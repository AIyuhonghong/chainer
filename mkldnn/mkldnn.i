%module mkldnn

%{
    #define SWIG_FILE_WITH_INIT
    #include "common.h"
    #include "convolution.h"
    #include "stream_factory.h"
%}

%include "numpy.i"

%init %{
    import_array();
    global_init();
%}

%apply ( float* IN_ARRAY4, int DIM1, int DIM2, int DIM3, int DIM4 )
    {( float* x, int x_d1, int x_d2, int x_d3, int x_d4 )}
%apply ( float* IN_ARRAY4, int DIM1, int DIM2, int DIM3, int DIM4 )
    {( float* gx, int gx_d1, int gx_d2, int gx_d3, int gx_d4 )}
%apply ( float* IN_ARRAY4, int DIM1, int DIM2, int DIM3, int DIM4 )
    {( float* W, int W_d1, int W_d2, int W_d3, int W_d4 )}
%apply ( float* IN_ARRAY4, int DIM1, int DIM2, int DIM3, int DIM4 )
    {( float* gW, int gW_d1, int gW_d2, int gW_d3, int gW_d4 )}
%apply ( float* IN_ARRAY1, int DIM1)
    {( float* b, int b_d1)}
%apply ( float* IN_ARRAY1, int DIM1)
    {( float* gb, int gb_d1)}
%apply ( float* INPLACE_ARRAY4, int DIM1, int DIM2, int DIM3, int DIM4 )
    {( float* y, int y_d1, int y_d2, int y_d3, int y_d4 )}
%apply ( float* INPLACE_ARRAY4, int DIM1, int DIM2, int DIM3, int DIM4 )
    {( float* gy, int gy_d1, int gy_d2, int gy_d3, int gy_d4 )}

%include "common.h"
%include "convolution.h"
%include "stream_factory.h"

%template(Convolution2D_F32) Convolution2D<float>;
