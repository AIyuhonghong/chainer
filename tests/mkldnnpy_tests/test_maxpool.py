import numpy as np
import math
from mkldnn import mkldnn
from chainer.utils import conv

n=1
c=1
h=4
w=4
stride=1
padding=1
ker=3
x = np.zeros((n, c, h, w), dtype=np.float32)
x2 = np.zeros((n, c, h, w), dtype=np.float32)
gy = np.zeros((n, c, h, w), dtype=np.float32)

for i in range(n):
    for j in range(c):
        for k in range(h):
            for l in range(w):
                x[i, j, k, l] = math.sin(i+j+k+l)
                x2[i, j, k, l] = math.sin(i+j+k+l)+1

for i in range(n):
    for j in range(c):
        for k in range(h):
            for l in range(w):
                gy[i, j, k, l] = 0.001;

print "x="
print x

y_h = conv.get_conv_outsize(h, k, stride, padding)
y_w = conv.get_conv_outsize(w, k, stride, padding)
y   = np.empty((n, c, y_h, y_w), dtype=x.dtype)
y2   = np.empty((n, c, y_h, y_w), dtype=x.dtype)
gx  = np.empty((n, c, h, w), dtype=x.dtype)
ws  = np.empty((n, c, y_h, y_w), dtype=np.int32)
ws2  = np.empty((n, c, y_h, y_w), dtype=np.int32)

mkldnn.MaxPooling_F32_do_forward(x, y, ws, stride, stride,
                    padding, padding, ker, ker);
print "y="
print y

mkldnn.MaxPooling_F32_do_forward(x2, y2, ws2, stride, stride,
                    padding, padding, ker, ker);
print "y2="
print y2
mkldnn.MaxPooling_F32_do_backward(gy, x, gx, ws, stride, stride,
                    padding, padding, ker, ker);
print "gx="
print gx
