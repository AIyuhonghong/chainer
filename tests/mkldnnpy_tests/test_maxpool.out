x=
[[[[ 0.          0.84147096  0.90929741  0.14112   ]
   [ 0.84147096  0.90929741  0.14112    -0.7568025 ]
   [ 0.90929741  0.14112    -0.7568025  -0.95892429]
   [ 0.14112    -0.7568025  -0.95892429 -0.27941549]]]]
y=
[[[[ 0.90929741  0.90929741  0.90929741  0.90929741]
   [ 0.90929741  0.90929741  0.90929741  0.90929741]
   [ 0.90929741  0.90929741  0.90929741  0.14112   ]
   [ 0.90929741  0.90929741  0.14112    -0.27941549]]]]
y2=
[[[[ 1.90929747  1.90929747  1.90929747  1.90929747]
   [ 1.90929747  1.90929747  1.90929747  1.90929747]
   [ 1.90929747  1.90929747  1.90929747  1.14111996]
   [ 1.90929747  1.90929747  1.14111996  0.72058451]]]]
gx=
[[[[ 0.     0.     0.006  0.   ]
   [ 0.     0.005  0.001  0.   ]
   [ 0.002  0.001  0.     0.   ]
   [ 0.     0.     0.     0.001]]]]
